PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterPlayerScripts/ControlScript/MasterControl/DynamicThumbstick.rbxmx

```diff
@@ -0,0 +1,415 @@
+
+local Thumbstick = {}
+
+--[[ Script Variables ]]--
+while not Players.LocalPlayer do
+	wait()
+end
+local LocalPlayer = Players.LocalPlayer
+
+local Backpack = LocalPlayer:WaitForChild("Backpack")
+local Tools = {}
+local IsAToolEquipped = false
+
+local IsPortraitMode = true
+
+local RevertAutoJumpEnabledToFalse = false
+
+local ThumbstickFrame = nil
+local ImageCenter = nil
+local ImageUnderThumb = nil
+local ImageRod = nil
+
+local IsFollowStick = false
+local ThumbstickFrame = nil
+local OnMoveTouchEnded = nil		-- defined in Create()
+local OnTouchMovedCn = nil
+local OnTouchEndedCn = nil
+local OnRenderSteppedCn
+local currentMoveVector = Vector3.new(0,0,0)
+
+--[[ Constants ]]--
+
+local IMAGE_OUTER_URL = "rbxasset://textures/ui/Disk.png"
+local IMAGE_STICK_URL = "rbxasset://textures/ui/Disk.png"
+local IMAGE_ROD_URL = "rbxasset://textures/ui/Square.png"
+
+local TOUCH_IS_TAP_TIME_THRESHOLD = 0.5
+local TOUCH_IS_TAP_DISTANCE_THRESHOLD = 25
+
+local FadeInAndOutBackground = true
+local FadeInAndOutMaxCount = nil
+local FadeInAndOutMaxAlpha = 0.35
+local FadeInAndOutHalfDuration = 0.3
+
+--[[ Local functionality ]]--
+
+local function adjustFrameSizeAndPosition()
+	if ThumbstickFrame then
+		if IsPortraitMode then
+			ThumbstickFrame.Size = UDim2.new(1, 0, .4, 0)
+			ThumbstickFrame.Position = UDim2.new(0, 0, 0.6, 0)
+		else
+			ThumbstickFrame.Size = UDim2.new(0.5, 0, 1, 0)
+			ThumbstickFrame.Position = UDim2.new(0, 0, 0, 0)
+		end
+	end
+end
+
+spawn(function()
+	while(true) do
+		--IsPortraitMode = newOrientation == Enum.ScreenOrientation.Portrait
+		wait()
+		local size = workspace.CurrentCamera.ViewportSize
+		local mode = size.X < size.Y
+		if mode ~= IsPortraitMode then
+			IsPortraitMode = mode
+			adjustFrameSizeAndPosition()
+		end
+	end
+end)
+
+local function onWorkspaceChanged(property)
+	if property == 'CurrentCamera' then
+		local newCamera = workspace.CurrentCamera
+		if newCamera then
+			local size = newCamera.ViewportSize
+			IsPortraitMode = size.X < size.Y
+			newCamera.Changed:Connect(function(property)
+				if property == 'ViewportSize' then
+					size = newCamera.ViewportSize
+					IsPortraitMode = size.X < size.Y
+					adjustFrameSizeAndPosition()
+				end
+			end)
+		end
+	end
+end
+
+local function setupTool(tool)
+	tool.Equipped:Connect(function()
+		IsAToolEquipped = true
+	end)
+	tool.Unequipped:Connect(function()
+		IsAToolEquipped = false
+	end)
+end
+	
+do
+	for _, item in ipairs(Backpack:GetChildren()) do
+		if item:IsA("Tool") then
+			Tools[#Tools + 1] = item
+			setupTool(item)
+		end
+	end
+	
+	Backpack.ChildAdded:Connect(function(item)
+		if item:IsA("Tool") then
+			setupTool(item)
+		end
+	end)
+	
+	workspace.Changed:Connect(onWorkspaceChanged)	
+	onWorkspaceChanged('CurrentCamera')
+end
+
+--[[ Public API ]]--
+function Thumbstick:Enable()
+	local humanoid = MasterControl:GetHumanoid()
+	if humanoid then
+		RevertAutoJumpEnabledToFalse = humanoid.AutoJumpEnabled == false
```

Comment:
I thought we were going to check DevTouchMovementMode == Enum.DevTouchMovementMode.UserChoice when DynamicThumbstick is enabled?